client/finality-grandpa/src/communication/gossip.rs

```diff
@@ -539,11 +539,13 @@ impl<N: Ord> Peers<N> {
 			Some(p) => p,
 		};
 
-		let invalid_change = peer.view.set_id > update.set_id ||
-			peer.view.round > update.round && peer.view.set_id == update.set_id ||
-			peer.view.last_commit.as_ref() > Some(&update.commit_finalized_height);
+		let valid_change = update.set_id > peer.view.set_id ||
+			update.set_id == peer.view.set_id && update.round > peer.view.round ||
+			update.set_id == peer.view.set_id &&
+				update.round == peer.view.round &&
+				Some(&update.commit_finalized_height) > peer.view.last_commit.as_ref();
```

Comment:
Maybe we could split this into multiple lines. 

```
let set_id_increased = update.set_id > peer.view.set_id;
let round_increased = update.set_id == peer.view.set_id && update.round > peer.view.round;
...
...
```

You see what I want.

Reply:
Is it better now? I've also added check that `commit_finalized_height` cannot move backwards event if set ID or round has increased.